guide/avoiding-unnecessary-configuration.md

```diff
@@ -0,0 +1,161 @@
+  <td align="center"><a href="../README.md#table-of-contents">Table of Contents</a></td>
+  <td align="right">Next: <a href="testing-and-running-your-plugins-locally.md">Testing and running your plugins</a></td>
+</tr></table>
+<!-- PreviousNext:END -->
+
+# Avoiding Unnecessary Configuration
+
+> Further reading: [Avoiding Unnecessary Task Configuration](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html)
+
+Every single Gradle build invocation must go through the [Configuration phase](lifecycle-of-a-gradle-build.md#configuration-phase). This is where the `build.gradle` scripts and plugins for project are evaluated and tasks are registered and configured, ready for Gradle to calculate dependencies and then execute the tasks. Any expensive actions in this phase will be run _every single time the user runs Gradle_ - as such we should make sure to avoid doing unnecessary work.
+
+The best way to avoid doing extra work is by making the realisation (when the objects are created) and configuration of `Task`s/`Configuration`s/other domain objects as _lazy_ as possible. This means that Gradle will not actually configure these objects unless it needs to. For example, if there's a task `foo` that we've lazily configured which isn't actually going to be run in the build - Gradle will not run the lazy configuration or even make ("realise") the `Task` object. The time savings may be small per task, but considering some repos have 1000s or 10,000s of tasks, the savings can be significant.
+
+## Lazy task registration
+
+When registering a task, you should use `register` instead of `create`:
+
+#### Bad:
+
+Using `create` here immediately _realises_ the `Task` object. The task is always configured (setting the input) even when the task is not going to be run.
+
+```java
+FooTask task = project.getTasks().create("foo", FooTask.class);
+task.getInput().set(something);
+```
+Any other lazy configuration of the task (eg in other plugins or scripts using `tasks).
+
+#### Good:
+
+Instead, we should use `register`:
+
+```java
+TaskProvider<FooTask> task = project.getTasks().register("foo", FooTask.class, foo -> {
+    foo.getInput().set(something);
+});
+```
+
+This is better; we tell Gradle there is task called `foo`, but we give the configuration of the task lazily. If `foo` is not executed, we don't even realise or configure the task.
+
+## Using `TaskProvider`s
+
+The `register` approach now gives us a `TaskProvider`, which is a lazy handle to the task. Many  of the built-in Gradle APIs will let you pass a `TaskProvider` instead of a `Task` object.
+
+If you want to give use an output property of a task as the input to another task, you can just transform the `TaskProvider` into another `Provider` using `map` or `flatMap`. You can call `.configure` on a `TaskProvider` to lazily configure the task without realising it (avoid calling `.get()` on `TaskProvider`s).
```

Comment:
I believe you always want to use `flatMap`. `TaskProvider<Provider<X>>` doesn't always carry dependencies iirc

Reply:
Yeah, I might leave that to the "task dependencies" section - or some other place about how task deps get removed. I'm also writing errorprones.

Reply:
>I believe you always want to use flatMap. TaskProvider<Provider<X>> doesn't always carry dependencies iirc